op.c

```diff
@@ -5011,6 +5011,38 @@ S_fold_constants(pTHX_ OP *const o)
         break;
     case OP_REPEAT:
         if (o->op_private & OPpREPEAT_DOLIST) goto nope;
+        /* Croak if the string is going to be unrealistically
+         * large. (GH#13324) Otherwise, don't constant fold
+         * above a certain threshold. (GH#13793 & GH#20586)
+         *
+         * Implementation note: pp_pow returns powers of 2 as an NV
+         *     e.g. my $x = "A" x (2**3);
+         */
+        if (OP_TYPE_IS(cBINOPo->op_last, OP_CONST)) {
+            SV *constsv = cSVOPx_sv(cBINOPo->op_last);
+            UV arbitrary = 1024 * 1024;
+
+            if (SvIOKp(constsv)) {
+                if (SvIOK_UV(constsv)) {
+                    if (SvUVX(constsv) > SIZE_MAX >> 2)
+                        ck_warner(packWARN(WARN_MISC), "Unrealistically large string repetition value");
+                    if (SvUVX(constsv) > arbitrary)
+                        goto nope;
+                } else {
+                    if (SvIVX(constsv) > (IV)(SIZE_MAX >> 2))
+                        ck_warner(packWARN(WARN_MISC), "Unrealistically large string repetition value");
+                    if (SvIVX(constsv) > (IV)arbitrary)
+                        goto nope;
+                }
+            } else {
+                NV rhs = 0.0; rhs = SvNV_nomg(constsv);
+                if (rhs >= (NV)((SIZE_MAX >> 2) +1) ) {
```

Comment:
should this be a -1.0 for safety? It can really questionable what the "53rd digit" to the right side of the `.` is, and what CC, what CPU, what OS, which security or spectre patch for your OS and CC, and make month and year of all 4 please.

I don't trust C's float/double keyword's rounding modes at all, and were constant folding intermediate values done in FP CPU real instructions or C abstract machine instructions, calculations done at 32, 64, or 80 bit or 128 bit intermediate floating point precision?

the goose has been cooked at malloc(2 GB) or malloc(2GB-1 byte) either way. thats not a future bug ticket.

Reply:
Also don't forget that Intel 64/AMD 64 in 64 bit mode CPUs are incapable of doing 80 bit floating pointer intermediate math unlike 32 bit mode. So  >= 2^53 or  >= 2^52 starts introducing more and more "error" or rounding into the math formula, and we have a 64 bit memory space  on paper (more like 48 bits unless your a rack server of brand new Xeons, which I think finally took another chomp at the AMD64 ISA's central address space gap).